### ISSUE TITLE
`make test` is failing in a new containerized environment

### ISSUE THREAD

**Issue by obedmr** [intel/pmem-CSI#153]
When trying to run basic validation with `make test`, it's failing in some dependencies missmatch. Below description of the taken steps.

Here's the `Dockerfile` I'm using to prepare the build dependencies
```
FROM golang:alpine

RUN apk add --update build-base autoconf automake bash-completion libtool libuuid json-c kmod asciidoc xmlto kmod-dev eudev-dev util-linux-dev json-c-dev linux-headers wget tar file diffutils git bash docker
RUN wget https://github.com/pmem/ndctl/archive/v63.tar.gz
RUN tar -xzvf v63.tar.gz
WORKDIR ndctl-63
RUN ./autogen.sh
RUN ./configure --libdir=/usr/lib --disable-docs --without-systemd
RUN make install
WORKDIR /go/src/github.com/intel/pmem-csi/
```
Then, on latest `devel` branch code.

**Build** - _SUCCESSFUL_
```
box ➜  pmem-csi git:(devel)  docker run -it  --rm -v `pwd`:/go/src/github.com/intel/pmem-csi pmem-csi-ci:go-alpine  make
GOOS=linux go build -a -o _output/pmem-csi-driver ./cmd/pmem-csi-driver
GOOS=linux go build -a -o _output/pmem-ns-init ./cmd/pmem-ns-init
GOOS=linux go build -a -o _output/pmem-vgm ./cmd/pmem-vgm
```

**Test** - _FAILED_
```
box ➜  pmem-csi git:(devel)  docker run -it  --rm -v `pwd`:/go/src/github.com/intel/pmem-csi pmem-csi-ci:go-alpine  make test
*** /dev/fd/63  Tue Jan 29 17:13:42 2019
--- /dev/fd/62  Tue Jan 29 17:13:42 2019
***************
*** 1,6 ****
  bitbucket.org/ww/goautoneg
- github.com/asaskevich/govalidator
  github.com/Azure/go-ansiterm
  github.com/beorn7/perks
  github.com/container-storage-interface/spec
  github.com/coreos/etcd
--- 1,10 ----
  bitbucket.org/ww/goautoneg
  github.com/Azure/go-ansiterm
+ github.com/NYTimes/gziphandler
+ github.com/PuerkitoBio/purell
+ github.com/PuerkitoBio/urlesc
+ github.com/Sirupsen/logrus
+ github.com/asaskevich/govalidator
  github.com/beorn7/perks
  github.com/container-storage-interface/spec
  github.com/coreos/etcd
***************
*** 15,27 ****
  github.com/evanphx/json-patch
  github.com/fatih/camelcase
  github.com/globalsign/mgo
- github.com/gogo/protobuf
- github.com/golang/groupcache
- github.com/golang/protobuf
- github.com/googleapis/gnostic
- github.com/google/btree
- github.com/google/gofuzz
- github.com/google/uuid
  github.com/go-openapi/analysis
  github.com/go-openapi/errors
  github.com/go-openapi/jsonpointer
--- 19,24 ----
***************
*** 32,37 ****
--- 29,41 ----
  github.com/go-openapi/strfmt
  github.com/go-openapi/swag
  github.com/go-openapi/validate
+ github.com/gogo/protobuf
+ github.com/golang/groupcache
+ github.com/golang/protobuf
+ github.com/google/btree
+ github.com/google/gofuzz
+ github.com/google/uuid
+ github.com/googleapis/gnostic
  github.com/gregjones/httpcache
  github.com/grpc-ecosystem/go-grpc-prometheus
  github.com/hashicorp/golang-lru
***************
*** 46,52 ****
  github.com/mitchellh/mapstructure
  github.com/modern-go/concurrent
  github.com/modern-go/reflect2
- github.com/NYTimes/gziphandler
  github.com/onsi/ginkgo
  github.com/onsi/gomega
  github.com/opencontainers/go-digest
--- 50,55 ----
***************
*** 58,66 ****
  github.com/prometheus/client_model
  github.com/prometheus/common
  github.com/prometheus/procfs
- github.com/PuerkitoBio/purell
- github.com/PuerkitoBio/urlesc
- github.com/Sirupsen/logrus
  github.com/spf13/cobra
  github.com/spf13/pflag
  golang.org/x/crypto
--- 61,66 ----
***************
*** 82,89 ****
  k8s.io/apiextensions-apiserver
  k8s.io/apimachinery
  k8s.io/apiserver
- k8s.io/client-go
  k8s.io/cli-runtime
  k8s.io/cloud-provider
  k8s.io/cluster-bootstrap
  k8s.io/csi-api
--- 82,89 ----
  k8s.io/apiextensions-apiserver
  k8s.io/apimachinery
  k8s.io/apiserver
  k8s.io/cli-runtime
+ k8s.io/client-go
  k8s.io/cloud-provider
  k8s.io/cluster-bootstrap
  k8s.io/csi-api

vendor-bom.csv not in sync with vendor directory (aka Gopk.lock):
+ new entry, missing in vendor-bom.csv
- obsolete entry in vendor-bom.csv
make: *** [test/test.make:28: test_vendor_bom] Error 1
```

I'm trying to understand more how dependencies are handled, I see you're using the `dep` tool's generated `Gopkg.*` files in conbination with some `*.csv` files.

Have you  considered migrating to the [modules](https://github.com/golang/go/wiki/Modules#migrating-to-modules) approach? It seems like it has fixed many of the `dep` tool issues.

**Comment by pohly** [comment]
Obed N Munoz <notifications@github.com> writes:
> box ➜  pmem-csi git:(devel)  docker run -it  --rm -v
> `pwd`:/go/src/github.com/intel/pmem-csi pmem-csi-ci:go-alpine  make
> test

Which revision are you testing?

> I'm trying to understand more how dependencies are handled, I see
> you're using the `dep` tool's generated `Gopkg.*` files in conbination
> with some `*.csv` files.

Dependencies are handled without those .csv files. We just manage those
on top of the underlying vendoring tool to comply with Intel processes.

> Have you considered migreating to the
> [modules](https://github.com/golang/go/wiki/Modules#migrating-to-modules)
> approach? It seems like it has fixed many of the `dep` tool issues.

I'm not sure yet which issues exactly that will solve.

**Comment by obedmr** [comment]
> Obed N Munoz <notifications@github.com> writes:
> box ➜ pmem-csi git:(devel) docker run -it --rm -v `pwd`:/go/src/github.com/intel/pmem-csi pmem-csi-ci:go-alpine make test
> Which revision are you testing?
latest in `devel`
> I'm trying to understand more how dependencies are handled, I see you're using the `dep` tool's generated `Gopkg.*` files in conbination with some `*.csv` files.
> Dependencies are handled without those .csv files. We just manage those on top of the underlying vendoring tool to comply with Intel processes.
> Have you considered migreating to the [modules](https://github.com/golang/go/wiki/Modules#migrating-to-modules) approach? It seems like it has fixed many of the `dep` tool issues.
> I'm not sure yet which issues exactly that will solve.

**Comment by obedmr** [comment]
@pohly, I'm using latest from `devel`

**Comment by obedmr** [comment]
@pohly  I tested it manually (in a new container) and in the CI and still the same issue. I tested it on my dev machine and it's actually running ok. I'd say that something is missing in the container to run it successfully. 

Could you validate if following steps are the ones required or if there's a missing one?

```
FROM golang:alpine

RUN apk add --update build-base autoconf automake bash-completion libtool libuuid json-c kmod asciidoc xmlto kmod-dev eudev-dev util-linux-dev json-c-dev linux-headers wget tar file diffutils git bash docker
RUN wget https://github.com/pmem/ndctl/archive/v63.tar.gz
RUN tar -xzvf v63.tar.gz
WORKDIR ndctl-63
RUN ./autogen.sh
RUN ./configure --libdir=/usr/lib --disable-docs --without-systemd
RUN make install
WORKDIR /go/src/github.com/intel/pmem-csi/
```

**Comment by obedmr** [comment]
and yes, I tested https://github.com/intel/pmem-CSI/commit/1665ee54d6cca92971cfaad723246bd730dd2ecf

**Comment by pohly** [comment]
Obed N Munoz <notifications@github.com> writes:
> ***************
> *** 82,89 ****
>   k8s.io/apiextensions-apiserver
>   k8s.io/apimachinery
>   k8s.io/apiserver
> - k8s.io/client-go
>   k8s.io/cli-runtime
>   k8s.io/cloud-provider
>   k8s.io/cluster-bootstrap
>   k8s.io/csi-api
> --- 82,89 ----
>   k8s.io/apiextensions-apiserver
>   k8s.io/apimachinery
>   k8s.io/apiserver
>   k8s.io/cli-runtime
> + k8s.io/client-go
>   k8s.io/cloud-provider
>   k8s.io/cluster-bootstrap
>   k8s.io/csi-api

Never mind, I think I understand what's failing: the version of sort
that is used inside the container sorts differently than the "normal"
sort, or the locale is different.

`LC_ALL=C LANG=C sort` might fix this.

**Comment by pohly** [comment]
"Patrick Ohly" <patrick.ohly@intel.com> writes:
> Never mind, I think I understand what's failing: the version of sort
> that is used inside the container sorts differently than the "normal"
> sort, or the locale is different.
>
> `LC_ALL=C LANG=C sort` might fix this.

Does it work on this branch? https://github.com/intel/pmem-CSI/tree/sort-order

**Comment by obedmr** [comment]
@pohly I tested your `sort-order` branch and it works.

**Comment by obedmr** [comment]
```
/go/src/github.com/intel/pmem-csi # make test
GOOS=linux go build -a -o _output/pmem-csi-driver ./cmd/pmem-csi-driver
TEST_WORK=/go/src/github.com/intel/pmem-csi/_work \
go test ./pkg/ndctl ./pkg/pmem-common ./pkg/pmem-csi-driver ./pkg/pmem-device-manager ./pkg/pmem-exec ./pkg/pmem-grpc ./pkg/pmem-registry
?       github.com/intel/pmem-csi/pkg/ndctl     [no test files]
?       github.com/intel/pmem-csi/pkg/pmem-common       [no test files]
ok      github.com/intel/pmem-csi/pkg/pmem-csi-driver   0.024s
?       github.com/intel/pmem-csi/pkg/pmem-device-manager       [no test files]
?       github.com/intel/pmem-csi/pkg/pmem-exec [no test files]
?       github.com/intel/pmem-csi/pkg/pmem-grpc [no test files]
?       github.com/intel/pmem-csi/pkg/pmem-registry     [no test files]
go vet github.com/intel/pmem-csi/pkg/...
/go/src/github.com/intel/pmem-csi #
```

**Comment by obedmr** [comment]
> "Patrick Ohly" <patrick.ohly@intel.com> writes:
> Never mind, I think I understand what's failing: the version of sort that is used inside the container sorts differently than the "normal" sort, or the locale is different. `LC_ALL=C LANG=C sort` might fix this.
> Does it work on this branch? https://github.com/intel/pmem-CSI/tree/sort-order

What's needed to have it merged? it would be really useful for the CI integration we're working on. Thanks @pohly

**Comment by pohly** [comment]
Fixed by PR #156.
